### ISSUE TITLE
Exception: We cannot tell the batch dim, when using RecLayer. 

### ISSUE THREAD

**Issue by jvhoffbauer** [rwth-i6/returnn#497]
When training an LM using the recurrent layer, I get the error: `Exception: We cannot tell the batch dim.` To me, this seems to be related to the usage of prev: in the RecLayer. 

I have attached a reproducer. 

[ner.tag_lm.config.txt](https://github.com/rwth-i6/returnn/files/6399177/ner.tag_lm.config.txt)
[language_models.py.txt](https://github.com/rwth-i6/returnn/files/6399179/language_models.py.txt) (sis pipeline, just for convenience)

**Comment by albertz** [comment]
Can you produce a simple test case (which no external dependencies)?

**Comment by jvhoffbauer** [comment]
Here you go

```

# start: nosetests $this_file --nologcapture

from __future__ import print_function

from pprint import pprint

import numpy.testing
import os

from returnn.config import Config
from returnn.tf.network import *
from returnn.tf.updater import Updater
from returnn.tf.util.basic import DimensionTag

tf_util.debug_register_better_repr()

from returnn.Dataset import init_dataset

import numpy as np

log.initialize(verbosity=[5])


# def make_feed_dict(network, step, seq_len=10, limit=1.0, random=numpy.random.RandomState(seed=1)):
#
#   d = {
#     network.extern_data.data["data"].placeholder: np.linspace(1, seq_len, seq_len).reshape(1, seq_len, 1), #random.uniform(-limit, limit, (1, seq_len, 1)),
#     network.extern_data.data["data"].size_placeholder[0]: numpy.array([seq_len]),
#     #network.extern_data.data["classes"].placeholder: random.uniform(-limit, limit, (1, seq_len, 1)),
#     #network.extern_data.data["classes"].size_placeholder[0]: numpy.array([seq_len]),
#   }
#   if isinstance(network.epoch_step, tf.Tensor):
#     d[network.epoch_step] = step
#   return d

def _check_train_simple_network():

  tf_compat.v1.set_random_seed(42)

  context_type = 'fixed'
  context_size = 2
  num_ner_labels = 45
  def language_model_encoder_subnet(from_name, num_hidden=64):
    if context_type == 'full':
      subnet = {
        'output': {
          'class': 'rnn_cell',
          'n_out': num_hidden,
          'unit': 'LSTMBlock'
        },
      }
    elif context_type == 'fixed':
      subnet = {
        'window': {
          'class': 'window',
          'window_size': context_size,
          'window_left': context_size - 1,
          'padding': 'same',
        },
        'merge': {
          'class': 'merge_dims',
          'axes': 'static',
          'from': ['window']
        },
        'output': {
          'class': 'linear',
          'from': ['merge'],
          'n_out': num_hidden,
          'activation': 'relu'
        },
      }
    else:
      assert False

    return {
      'class': 'subnetwork',
      'from': [from_name],
      'subnetwork': subnet
    }

  network = {
    'output': {
      'class': 'rec',
      'from': 'data:classes',
      'unit': {

        'tag_embedding': {
          'class': 'linear',
          'activation': None,
          'from': ['data:source'],
          'n_out': 256,
          'with_bias': True,
          'initial_output': 'zeros',
          # 'initial_input': 'zeros',
          # 'initial_state': 'zeros',
        },

        'hidden': language_model_encoder_subnet('prev:tag_embedding'),

        'output': {
          'class': 'linear',
          'from': ['hidden'],
          'n_out': num_ner_labels,
          'activation': 'softmax',
          'loss': 'ce',
          'target': 'classes'
        },
      }
    }
  }

  dec_time = DimensionTag(kind=DimensionTag.Types.Spatial, description="combined_time")

  config = Config()
  config.update({
    #"num_inputs": 1,
    #"num_outputs": {"data": [1, 2], "classes": [1, 2]},  # dense output
    "network": network,
    "adam": True,
    #'optimizer': "adam",
    "learning_rate": 0.1,
    "debug_add_check_numerics_ops": True,
    "extern_data": {
      'classes': {
        'dim': num_ner_labels,
        'sparse': True,
        'same_dim_tags_as': {'t': dec_time},
      },
    },
  })

  # p(c_n | c_n-1, c_n-2)
  # <sb> Sam walks on the road

  with tf.Graph().as_default() as graph:
    with tf_compat.v1.Session(graph=graph) as session:

      # Create network from config
      print("Create network...")
      network = TFNetwork(config=config, train_flag=True)
      network.construct_from_dict(config.typed_dict["network"])
      network.print_network_info()
      #network.initialize_params(session=session)

      # # Get fetches from network
      # fetches = network.get_fetches_dict()
      # pprint(fetches)
      #
      # # Initialize variables in network
      # session.run(tf_compat.v1.variables_initializer(tf_compat.v1.global_variables() + [network.global_train_step]))
      #
      # # Prepare input
      # #data_input = network.extern_data.data["data"]
      # #rand_data = rnd.rand(n_batch, n_time, n_out)
      #
      # # Create updater
      # updater = Updater(config=config, network=network)
      # updater.set_learning_rate(config.float("learning_rate", 0.1), session=session)
      # updater.set_trainable_vars(network.get_trainable_params())
      # updater.init_optimizer_vars(session=session)
      #
      # # Run training
      # num_steps = 100
      # for step in range(num_steps):
      #   result = session.run(
      #     fetches=[
      #       network.get_total_loss(),
      #       network.get_layer("output").output.placeholder,
      #       updater.get_optim_op(),
      #       network.get_post_control_dependencies()
      #     ],
      #     feed_dict=make_feed_dict(network, 1, 10)
      #   )
      #   print("step %i, loss: %r" % (step, result))
      #
      # # Run network
      # info, out = session.run(
      #   fetches=[
      #     fetches,
      #     network.get_layer("output").output.placeholder,
      #   ],
      #   feed_dict=make_feed_dict(network, 1, 10),
      #   # feed_dict={
      #   #   data_input.placeholder: rand_data,
      #   #   data_input.size_placeholder[0]: numpy.array([n_time] * n_batch, dtype="float32"),
      #   # })
      # )
      #
      # pprint(info)
      # pprint(out)
      # print(out.shape)

# def create_dataset():
#
#   dataset_path = '/u/hoffbauer/code/returnn-ner-setup/work/ner/corpus/ConvertCoNLLToRETURNNFormat.6Rq4FReWf6af/output/corpus'
#   # split = 'train'
#
#   train_ner = {
#     'class': 'TranslationDataset',
#     'file_postfix': 'train',
#     'partition_epoch': 1,
#     'path': dataset_path,
#     'seq_ordering': 'random',
#     'source_postfix': '',
#     'target_postfix': '',
#     'unknown_label': '<UNK>'
#   }
#
#   train_data = init_dataset(
#     train_ner
#   )
#
#   pprint(train_data)


if __name__ == '__main__':
    _check_train_simple_network()
```

**Comment by albertz** [comment]
By "simple", I mean as short as possible. Can you clean that up?

**Comment by jvhoffbauer** [comment]
So I've uploaded a simplified version.

**Comment by albertz** [comment]
Ah I see. The problem is that you simply have `classes` here, and it has set `available_for_inference = False`.

Use:
```
      'classes': {
        'dim': num_ner_labels,
        'sparse': True,
        'same_dim_tags_as': {'t': dec_time},
        'available_for_inference': False
      },
```

(Btw, it would be good if you also try to debug things like these yourself.)
